### DISCUSSION TITLE
Self-Hosted Github: Release Notes Missing for Python Packages

### DISCUSSION THREAD

**Discussion by scemily13** [renovatebot/renovate#15604] [Request Help]
Hello!

We've been running self-hosted Renovate with Github and have not been getting release notes for just about all Python packages. An example below:
<img width="842" alt="Screen Shot 2022-05-16 at 12 31 29 PM" src="https://user-images.githubusercontent.com/68038396/168640394-03271056-194c-432b-8220-08e518077882.png">

This does not appear to be an issue for other language dependencies:
<img width="851" alt="Screen Shot 2022-05-16 at 12 32 27 PM" src="https://user-images.githubusercontent.com/68038396/168640610-5935387e-4896-4e74-8519-9a6011dc3ab9.png">

We are using the `GITHUB_COM_TOKEN`, so have ruled that out as a potential cause.

**Comment by rarkins** [comment]
For release notes to be shown, the registry must return the source repository in metadata. It's possible GitLab does not do this for Python packages.

The logic for Renovate finding the `sourceUrl` field for python is here: https://github.com/renovatebot/renovate/blob/da9b94de84bf058de330451d0e4c1b09302b5b0d/lib/modules/datasource/pypi/index.ts#L104-L126

You would need to do something like use `curl` to query an example package on your GitLab and look at what it returns to see if the github.com source URLs you're expecting are anywhere there which Renovate could find.

**Reply by scemily13** [reply]
So we use Github, not GitLab, but discovered that the likely culprit here is the fact that our Artifactory mirror of PyPI [does not support the PyPI JSON API](https://www.jfrog.com/jira/browse/RTFACT-16101). Thus Renovate queries our `/simple` endpoint where the corresponding Github repository is not listed.

**Reply by scemily13** [reply]
Would y'all be open to allowing an alternative path for retrieving Python dependency changelogs, that looks for entries in the Github `/releases` url or in a CHANGELOG file for a repo?

**Reply by rarkins** [reply]
@scemily13 how would you propose that Renovate can know which GitHub repo to query the `/releases` from though? Isn't the problem that the source URL for packages is missing from `/simple` and therefore Renovate doesn't know where to start?

**Reply by scemily13** [reply]
@rarkins There are a few different avenues Renovate could take, but one would be just grabbing the `Homepage` link from the [PyPI page](https://pypi.org/project/ddtrace/), under 'Project links' as this often contains the Github repo.

**Reply by rarkins** [reply]
Why can't we just look up versions from pypi in the first place and ignore artifactory altogether?

**Reply by scemily13** [reply]
That would work!

**Reply by rarkins** [reply]
How/where is the artifactory url being learned? You should be able to override it by specifying registryUrls pointing to pypi for all packages which aren't private

**Reply by scemily13** [reply]
We have some security concerns at the moment that prevent us from including the public PyPI URL in our `registryUrls`. That being said we might be able to take advantage of include/exclude package patterns in this case.
